pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+		processor.ProvisionTypeConfig,
+	)
+	if err != nil {
+		return cerrors.Errorf("failed to create processor: %w", err)
+	}
+	return nil
+}
+func (a createProcessorAction) Rollback(ctx context.Context) error {
+	err := a.processorService.Delete(ctx, a.cfg.ID)
+	if !cerrors.Is(err, processor.ErrInstanceNotFound) {
+		// ignore instance not found errors, this means the action failed to
+		// create the processor in the first place
+		return err
+	}
+	return nil
+}
+
+// --------------------
+// -- UPDATE ACTIONS --
+// --------------------
+
+type updatePipelineAction struct {
+	oldConfig config.Pipeline
+	newConfig config.Pipeline
+
+	pipelineService PipelineService
+}
+
+func (a updatePipelineAction) String() string {
+	return fmt.Sprintf("update pipeline with ID %v", a.oldConfig.ID)
+}
+func (a updatePipelineAction) Do(ctx context.Context) error {
+	return a.update(ctx, a.newConfig)
+}
+func (a updatePipelineAction) Rollback(ctx context.Context) error {
+	return a.update(ctx, a.oldConfig)
+}
+func (a updatePipelineAction) update(ctx context.Context, cfg config.Pipeline) error {
+	p, err := a.pipelineService.Update(ctx, cfg.ID, pipeline.Config{
+		Name:        cfg.Name,
+		Description: cfg.Description,
+	})
+	if err != nil {
+		return cerrors.Errorf("failed to update pipeline: %w", err)
+	}
+	_, err = a.pipelineService.UpdateDLQ(ctx, cfg.ID, pipeline.DLQ{
+		Plugin:              cfg.DLQ.Plugin,
+		Settings:            cfg.DLQ.Settings,
+		WindowSize:          *cfg.DLQ.WindowSize,
+		WindowNackThreshold: *cfg.DLQ.WindowNackThreshold,
+	})
+	if err != nil {
+		return cerrors.Errorf("failed to update pipeline DLQ: %w", err)
+	}
+
+	// update connector IDs
+	if !a.isEqualConnectors(p.ConnectorIDs, cfg.Connectors) {
+		// recreate all connector IDs
+		for _, procID := range p.ConnectorIDs {
+			_, err = a.pipelineService.RemoveConnector(ctx, cfg.ID, procID)
+			if err != nil {
+				return cerrors.Errorf("failed to remove connector %v: %w", procID, err)
+			}
+		}
+		for _, proc := range cfg.Connectors {
+			_, err = a.pipelineService.AddConnector(ctx, cfg.ID, proc.ID)
+			if err != nil {
+				return cerrors.Errorf("failed to add connector %v: %w", proc.ID, err)
+			}
+		}
+	}
+
+	// update processor IDs
+	if !a.isEqualProcessors(p.ProcessorIDs, cfg.Processors) {
+		// recreate all processor IDs
+		for _, procID := range p.ProcessorIDs {
+			_, err = a.pipelineService.RemoveProcessor(ctx, cfg.ID, procID)
+			if err != nil {
+				return cerrors.Errorf("failed to remove processor %v: %w", procID, err)
+			}
+		}
+		for _, proc := range cfg.Processors {
+			_, err = a.pipelineService.AddProcessor(ctx, cfg.ID, proc.ID)
+			if err != nil {
+				return cerrors.Errorf("failed to add processor %v: %w", proc.ID, err)
+			}
+		}
+	}
+
+	return nil
+}
+func (updatePipelineAction) isEqualConnectors(ids []string, connectors []config.Connector) bool {
+	if len(ids) != len(connectors) {
+		return false
+	}
+	for i := range ids {
+		if ids[i] != connectors[i].ID {
```

Comment:
i.e. ordering of the IDs and connectors is important?

Reply:
Yes, I thought that since we now have deterministic ordering of both connectors and processors that we can make sure the IDs really are stored in the specified order. Note that this doesn't mean that the connectors will be updated if you change their order in the config file, only the `ConnectorIDs` and `ProcessorIDs` fields in the pipeline will be updated.